src/EFCore.SqlServer/Migrations/SqlServerMigrationsSqlGenerator.cs

```diff
@@ -3118,7 +3118,12 @@ alterTableOperation.OldTable[SqlServerAnnotationNames.TemporalHistoryTableSchema
                         var changeToSparse = alterColumnOperation.OldColumn[SqlServerAnnotationNames.Sparse] as bool? != true
                             && alterColumnOperation[SqlServerAnnotationNames.Sparse] as bool? == true;
 
-                        if (changeToNonNullable || changeToSparse)
+                        // for alter column removing default value we also need to disable versioning
+                        // because the default constraint needs to be removed from both main and history tables
+                        var removingDefaultValue = (alterColumnOperation.OldColumn.DefaultValue != null || alterColumnOperation.OldColumn.DefaultValueSql != null)
+                            && alterColumnOperation.DefaultValue == null && alterColumnOperation.DefaultValueSql == null;
```

Comment:
Specifically since DefaultValue is object and can be anything (and so may override the equality operator etc).

```suggestion
                        var removingDefaultValue = (alterColumnOperation.OldColumn.DefaultValue is not null || alterColumnOperation.OldColumn.DefaultValueSql is not null)
                            && alterColumnOperation.DefaultValue is null && alterColumnOperation.DefaultValueSql is null;
```